### PR TITLE
Fix #3818: Datatable unsort support

### LINKED ISSUE #3818: DataTable: add ability to unsort column
At the moment, 6.2.5, a user cannot 'unsort' a column, only asc and desc are offered.

This in combination with the nice multiViewState feature forces a logout to undo sorting.

### PR DESCRIPTION
**Pull request by eduarddrenth**
- SortEvent supports all SortOrder
- unsort support in datatable.js
- wrong aria message keys
- docs

### CODE REVIEW

**Review by melloware — APPROVED** [review]

**Comment by tandraschko** [comment]
WDYT @mertsincan?

**Comment by eduarddrenth** [comment]
Note that this PR may cause work-arounds (like the one I had myself) to fail.

**Comment by tandraschko** [comment]
@eduarddrenth i have the GO from PrimeTek.
See https://primefaces.org/primereact/showcase/#/datatable/sort

Should we also add a removableSort flag?

**Comment by eduarddrenth** [comment]
When removableSort==false don't unsort? For me it is pretty intuitive without it, anyone else?

**Comment by tandraschko** [comment]
in older PrimeFaces it was not easily possible to unsort. So maybe users like to restore the old behavior which dont allow unsort?

**Comment by eduarddrenth** [comment]
Perhaps, to be backwards compatible the default should than be false. Such an option requires a bit of coding and review. Break open the PR? New issue? More discussion?

**Comment by melloware** [comment]
No we should make the new one unsort by default but give the user the ability to turn it off like @tandraschko said.  The new default is a good one.

**Comment by tandraschko** [comment]
yep
i would call the attribute "allowUnsorting"

File: src/main/resources/META-INF/resources/primefaces/datatable/datatable.js [hunk 1]
```diff
@@ -530,6 +531,7 @@ PrimeFaces.widget.DataTable = PrimeFaces.widget.DeferredWidget.extend({
 
             var columnHeader = $(this),
             sortOrderData = columnHeader.data('sortorder'),
+                // TODO here we should exclude unsorted based on cfg
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/datatable/datatable.js]: Is the TODO meaning you still need to do it?

Comment (eduarddrenth) [reply src/main/resources/META-INF/resources/primefaces/datatable/datatable.js]: Yes, still working on it in my fork

**Review by melloware — APPROVED** [review]

**Comment by eduarddrenth** [comment]
Added below to Primefaces.d.ts for cfg, did not test if everything works:
```
    export interface DataTableCfg extends PrimeFaces.widget.DeferredWidgetCfg {
        /**
         * Default is true: include "unsort" when repeatedly clicking on column header.
         */
        allowUnsorting: boolean;
```

**Comment by eduarddrenth** [comment]
Testing...I get a console error, fixing datatable.js

**Comment by eduarddrenth** [comment]
tested, working, option is working too

File: src/main/resources/META-INF/primefaces-p.taglib.xml [hunk 2]
```diff
@@ -7300,6 +7300,14 @@
             <required>false</required>
             <type>java.lang.String</type>
         </attribute>
+        <attribute>
+            <description>
+                <![CDATA[Defines wether columns are allowUnsorting.]]>
```
Comment (melloware) [inline src/main/resources/META-INF/primefaces-p.taglib.xml:7305]: Defines whether columns are are allowed to be unsorted. Default is true.

File: docs/9_0/components/datatable.md [hunk 3]
```diff
@@ -106,6 +106,7 @@ DataTable displays data in tabular format.
 | tabindex                  | null               | String           | Position of the element in the tabbing order.
 | tableStyle                | null               | String           | Inline style of the table element.
 | tableStyleClass           | null               | String           | Style class of the table element.
+| allowUnsorting                | true               | Boolean          | Include "unsort" when repeatedly clicking on column header.
```
Comment (melloware) [inline docs/9_0/components/datatable.md]: Defines whether columns are are allowed to be unsorted. Default is true.

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by eduarddrenth** [comment]
changed text

**Review by melloware — APPROVED** [review]

**Comment by tandraschko** [comment]
thanks, its a great contribution!

**Comment by eduarddrenth** [comment]
on to the next.... "retain order of sort columns with multi column sort and multiviewstate"

**Comment by Rapster** [comment]
Are you sure you want to make allowUnSorting to true by default? As I'm using sorting a lot lately, I must say it's confusing whenever i click on column's header the result does not change while I expect it to be sorted asc or desc. I quite agree with Thomas statement:
> in older PrimeFaces it was not easily possible to unsort. So maybe users like to restore the old behavior which dont allow unsort?

Because it was not possible before, and it's more intuitive to have this transition: `UNSORTED > ASC > DESC > ASC > DESC`, I would put this property to false by default, WDYT?

**Comment by eduarddrenth** [comment]
Changes can be confusing to people yes, but for me uns->asc->desc->uns is more intuitive.

Anyway, when allowUnSorting is false by default there will be no (expected) complaints and the complaint
"I cannot unsort" can be dealt with.

**Comment by Rapster** [comment]
> Changes can be confusing to people yes, but for me uns->asc->desc->uns is more intuitive.

I'm not talking changes in general, but this one specifically. I think most users expect the scenario I mentionned, you can see few examples [here](https://mdbootstrap.com/docs/jquery/tables/sort/) and [there](https://datatables.net/), and excel does almost the same

> Anyway, when allowUnSorting is false by default there will be no (expected) complaints and the complaint
"I cannot unsort" can be dealt with.

Exactly, and the idea is to have as less people complaining as possible (right? 😅) so allowUnSorting to false by default makes more sense no?

**Comment by stolp** [comment]
It definitely makes sense not to change the default behavior. As a framework consumer I hate surprising changes without my consent.

**Comment by melloware** [comment]
So are we fixing this? I agree with @stolp and @Rapster

**Comment by Rapster** [comment]
It's in my todo list for #5438 😉

**Comment by eduarddrenth** [comment]
way to go

**Pull request merged**